### PR TITLE
Set Up Funtions

### PR DESCRIPTION
**Pull request by mem48**
A new file otp-setup.R containing 4 new functions

otp_build_graph - builds an OPT graph
otp_setup - starts a local OPT server
otp_stop - stops a local OPT server 
otp_checks - runs some basic checks, do we have the correct version of Java installed etc. Called internally by other above functions

The idea is that people who are R users may not be familiar with Java and/or command line so we provide these functions to simplify setting up the OTP.

Currently only tested on Windows 10, should work on other windows version but some functions like otp_stop will not work on other OS.

Simple example:
```
# Create folder "C:/otp" and fill with otp.jar , graphs folder etc
# In R run
otp_build_graph("C:/otp") # Will add the Graph.obj to "C:/otp/graphs/current" folder
otp_setup("C:/otp") # Will run OTP (check with browser)
otp_stop() # Will Kill Java
```

### CODE REVIEW

**Comment by mem48** [comment]
P.S. I haven't rebuilt the descriptions etc to avoid merge conflicts

**Comment by Robinlovelace** [comment]
Any thoughts on this @marcusyoung will be good to get it rolling.

**Comment by marcusyoung** [comment]
I didn't get a notification from Github about this. I'll take a look.

**Review by marcusyoung — CHANGES_REQUESTED** [review]
@mem48 

Just an initial thing I spotted. otp-connect has a make_url method which generates the baseurl, so that just needs to be called at line 63.

I will try and look at all your other good stuff as soon as I can.

**Comment by mem48** [comment]
@marcusyoung I've switched to make_url() and also made a load more improvements.

1. Now have the otp_route() function which is working but would welcome some testing
2. otp_setup will now check the connection (using otp_connect) and if successful will open OTP in the browser
3. Fixed a bug in check_router.otpconnect() where the code fails rather than returning the expected message.

**Comment by mem48** [comment]
I've made some general updates to make it more "package friendly". Adding some vignettes based on your tutorial and sample data. I've also updated the description and documentation so `devtoolls::install_github()` works.

**Comment by Robinlovelace** [comment]
This PR is getting huge! Looks good, suggest merging soon. Can always tidy-up details after the event.

Biggest problem from CRAN perspective: giant size. 30 MB+. Suggest using the piggyback package or similar to help separate code and data:

```
git clone git@github.com:mem48/opentripplanner-malcolm
Cloning into 'opentripplanner-malcolm'...
remote: Counting objects: 291, done.
remote: Compressing objects: 100% (147/147), done.
remote: Total 291 (delta 114), reused 172 (delta 66), pack-reused 70
Receiving objects: 100% (291/291), 34.04 MiB | 5.79 MiB/s, done.
Resolving deltas: 100% (131/131), done.
```

**Comment by mem48** [comment]
The large size is the sample data, I'd been copying @marcusyoung tutorial.

I slimmed down his original data a bit but still

DEM.tif 5 MB
greater-manchester-osm.pbf 11 MB
tfgm-gtfs.zip 33 MB

Perhaps we should build a new tutorial on a smaller location, e.g. the Isle of White. The problem then will be GTFS data which is waiting on my https://github.com/mem48/UK2GTFS package

File: R/otp-setup.R
Comment (marcusyoung) [inline R/otp-setup.R:36]: @mem48 r.e. `otp_build_graph()`

For the help example, as a minimum need to provide:

`otp_build_graph(otp = "C:/otp", dir = "C:/otp")`

If the JAR file is also at the dir location, would it be nice to only have to specify that, and not otp too?

There's an issue with reaching max.print. As the build process tends to produce many  warnings. So I never got to see the message that graph build was ok. Would it be better to pipe the OTP build output to a log file instead?

**Comment by marcusyoung** [comment]
@mem48 @Robinlovelace 
Apologies for delay with this. Now reviewing. I'll comment as I go, but am not saying these issues all need addressing before I merge. As Robin said these can be tidied up later if needs be. One general thing is there's a lot in this pull request. Probably be best to have a separate branch per function/group of related functions in future if there aren't dependencies between? Makes it easier to review and quicker to merge as an issue in one area won't hold up those ready to go.

File: R/otp-setup.R
Comment (marcusyoung) [inline R/otp-setup.R]: I think this message should be \"Failed to start OTP\" or something along those lines.

File: R/otp-setup.R
Comment (marcusyoung) [inline R/otp-setup.R:157]: @mem48 `otp_build_graph()`
I think this message should be \"Failed to start OTP\" or something along those lines.

File: R/otp-setup.R
Comment (marcusyoung) [inline R/otp-setup.R:174]: @mem48 Best to use the `make_url` method from `otp_connect()` here.

File: R/otp-setup.R
Comment (marcusyoung) [inline R/otp-setup.R]: @mem48
```
if(!exists("otp")){
    warning("Path to the Open Trip Planner is not defined")
    stop()
  }
  if(!exists("dir")){
    warning("Path to the files to build graph are not defined")
    stop()
  }
```

These checks aren't doing as expected so R throws error if either are not provided. otp and dir will always exist as `otp = null` etc creates an empty object. Want to be checking !is.null() ?

File: R/otp-setup.R
Comment (marcusyoung) [inline R/otp-setup.R:249]: @mem48 AFAIK JAVA version must be 1.8. It won't work on 9 or 10.

File: R/otp-setup.R
Comment (marcusyoung) [inline R/otp-setup.R:260]: @mem48 I've posted a general issue re standardising on using checkmate package for argument checks. It includes file IO and OS checks. And multiple checks can be grouped. Would really make the code much more concise. E.g. see use in `otp_connect()`.

**Comment by marcusyoung** [comment]
@mem48  Just finished reviewing otp-setup.R. Some great stuff in there. I've made a few comments, see above. Quite a few typos in the help content, not worrying about that at the moment, but will need checking thoroughly at some point.

File: R/otp-route.R
Comment (marcusyoung) [inline R/otp-route.R]: @mem48 Check of modes needs to be a bit more nuanced. Some combinations of these are invalid. Should probably also allow lower or uppercase and do an upper conversion in the script. Don't want to throw errors when we know what the user means and can correct?

File: R/otp-route.R
Comment (marcusyoung) [inline R/otp-route.R:94]: @mem48 Need to deal with the \'walkTime\' label bug, where it is used for cycle time and drive time too. See line 162 etc. in my original function (in R scripts folder). Though this needs some thought as, for example, if a cycle route involves some walking, everything will be summed under `\'walkTime\'. Same issue with \'walkDistance\'.

File: R/otp-route.R
Comment (marcusyoung) [inline R/otp-route.R:1]: @mem48 This is actually returning other variables related to the route, and not just the geometry.  I guess this could just be an option on `trip_plan()`. By default don't return the geometry (wouldn't want the processing overhead if not required).  Is a separate `otp_trip_geometry()` function still useful, that just returned the geometry and not the other data?

File: R/otp-batch.R
Comment (marcusyoung) [inline R/otp-batch.R:1]: @mem48 I'm not sure this is the way to go with batch processing. This is likely to be a requirement of many of the functions, so it would be better if there was a single internal function which the other functions can call. The user should be able to provide either a single from/to set of coordinates OR a matrix -  in all relevant functions. Otherwise there will be a lot of repetition. Suggest moving this out of the pull request for now or at least not exporting any functions from it. What do you think?

File: R/otp-route.R
Comment (marcusyoung) [inline R/otp-route.R:21]: @mem48 Note typo: `@detials`

File: R/otp-route.R
Comment (marcusyoung) [inline R/otp-route.R:2]: @mem48 I think we probably need to make clear in the sf column names what the units are for duration, time, distance, elevationGained etc. E.g change labels to distance_km, time_seconds etc. Rather than relying on the user knowing?

File: R/otp-route.R
Comment (marcusyoung) [inline R/otp-route.R]: @mem248 Having problems getting TRANSIT (or BUS) route returned for a specific date. For example, the following on OTP with my Manchester tutorial data, returns trip is not possible error but a valid route is returned using the web GUI, where otpdate is `"2017-10-10 12:15:00 BST"`:

```
otp_plan(otpcon, from = c(53.56172, -2.29202), to = c(53.40892, -2.16637), mode = "BUS", full_elevation = FALSE, date_time = otpdate)
```

This may be due to the date being in the format YYYY-MM-DD when it's extracted from the POSIXct object, while OTP requires MM-DD-YYYY? On that point, I wonder if we really want to require users to provide date and time in POSIXct? It's not that user friendly, and many won't know how to create the object. I had to Google it to remind myself. Think I'd rather provide a basic text format option (or the option of either). What do you think?

File: R/otp-route.R
Comment (marcusyoung) [inline R/otp-route.R:109]: Is Google relevant here?

**Comment by marcusyoung** [comment]
@mem48 I've reviewed otp-route. Comments on specific bits of code above. Basic routing by car/walk/cycle seems to be working great. There's a problem requesting TRANSIT/BUS/TRAM/RAIL routes as I've mentioned above that needs looking into. I think it is being caused by the date format YYYY-MM-DD instead of MM-DD-YYYY required by OTP. Lots of typos to look at, including in user-facing stuff - help and error messages etc.

File: R/otp-geocode.R
Comment (marcusyoung) [inline R/otp-geocode.R:28]: type argument not specified. Function fails.

**Comment by marcusyoung** [comment]
@mem48 Taken a look at otp-geocode. The type argument is missing from the function definition so unable to test.

File: R/otp-isochrone.R
Comment (marcusyoung) [inline R/otp-isochrone.R:31]: @mem48 I don't think CAR should be the default here as it isn't a valid mode for isochrones. It will return polygons but they won't be correct. Need to think about whether we even allow non-transit modes as options here.

**Comment by marcusyoung** [comment]
@mem48 Have taken a look at otp-isochrone. This has similar problem to otp-route for TRANSIT modes when specifiying date and time, I think because the date format is not right for OTP. Currently returns empty polygons in this situation, so also need to look at how we pick up errors.

Function currently returns only an SF object. I think going forward the return format should be optional, additionally including GEOJSON (and possibly SHAPEFILE) Both are the native options available from OTP. Users will not always want to continue processing in R, e.g. writing direct to PostgreSQL.

File: vignettes/dummies_guide.Rmd
Comment (marcusyoung) [inline vignettes/dummies_guide.Rmd:57]: @mem48 Probably need a word of warning here. The release version of OTP is not well-suited for all counries (it is tuned for USA). The OTP.JAR in the tutorial is a modified version where I change edge traversal permissions and adjusted speed limits to better match the UK situation. I think that this may now be possible to configure via the new Way property sets that can be configured and are then read at graph build. I will investigate further. This info can then be incorporated into the advanced guide (probably?).

File: vignettes/advanced_features.Rmd
Comment (marcusyoung) [inline vignettes/advanced_features.Rmd:63]: @mem48 There's something not right about these isochrones in the map. Full of holes. Compare with the one in the tutorial. It may due to WALK being included in the modes. Isochrones only work correctly for TRANSIT modes.

**Comment by marcusyoung** [comment]
@mem48 Have had a look at the new Vignettes. Great work. A few comments above.

**Comment by marcusyoung** [comment]
@mem48 Looks like we've got some internal functions appearing in the man folder.

File: R/otp-setup.R
Comment (mem48) [reply R/otp-setup.R:36]: I went back and forth on having a separate otp and dir location. On balance I think a separation of code and data is preferable. If a user really wants them in the same place then having to complete two variables is a small hurdle. The change I have made is to make `otp` a path to the file not the folder, this avoids the problem of a folder containing more than one jar file and makes the difference between the two variables clearer.

I also change the help example to show that the results of the log can be stored

File: R/otp-setup.R
Comment (mem48) [reply R/otp-setup.R:157]: Not always true, OTP may have started but the build process failed. e.g. missing/corrupt data. So I felt that failed to build graph was more informative.

File: R/otp-setup.R
Comment (marcusyoung) [reply R/otp-setup.R:36]: @mem48 OK, that will make the difference between the two clearer.

File: R/otp-setup.R
Comment (marcusyoung) [reply R/otp-setup.R:157]: @mem48  Sorry I have put the comment in the wrong place. I was referring to `otp_setup()` where the message is the same.

File: R/otp-setup.R
Comment (mem48) [reply R/otp-setup.R:174]: good call

File: R/otp-setup.R
Comment (mem48) [reply R/otp-setup.R:249]: According to http://docs.opentripplanner.org/en/latest/Basic-Tutorial/#start-up-otp it is 1.8 or higher. I'm cautious of being too prescriptive as it will mean updating the package each time OTP or Java get an update. It was more intended as a test that the R to Java communication was working.

Comment (marcusyoung) [reply R/otp-setup.R:249]: That tutorial is incorrect. See issue raised by Robin. Get your point about being too prescriptive but OTP is likely to be very slow in updating for new Java versions. Maybe more important in the Vignettes that it is clear otherwise likely to cause huge frustration.

File: R/otp-setup.R
Comment (mem48) [reply R/otp-setup.R:260]: done

File: R/otp-route.R
Comment (mem48) [reply R/otp-route.R:94]: I would not change this, this is an OTP bug and should be fixed in the OTP, if we start making these changes it will become a nightmare supporting different versions of OTP and moves the package away from being a simple bridge between OTP and R

Comment (marcusyoung) [reply R/otp-route.R:94]: Agreed. I will submit bug report on this to OTP (has been around a long time).

File: R/otp-route.R
Comment (mem48) [reply R/otp-route.R:1]: The function basically gets everything that OTP returns, that seems a good default. We could add a variable to only return the metadata, which would provide a small performance boost

Comment (marcusyoung) [reply R/otp-route.R:1]: That would be useful I think.

File: R/otp-setup.R
Comment (mem48) [reply R/otp-setup.R:249]: Ok, changed to check that version is 8.X

File: R/otp-route.R
Comment (mem48) [reply R/otp-route.R]: uppercase text is easy, what are the rules on which combinations are allowed?

File: R/otp-route.R
Comment (mem48) [reply R/otp-route.R:1]: ok I will add

File: R/otp-route.R
Comment (marcusyoung) [reply R/otp-route.R]: I'll probably need to dig into the code to get definite answer on that one, though there may be something in the Google Group. It may be best to stick with the options available in the GUI and the modes they generate in the URL (after clicking the view this itinerary link (or whetever it's called). Leave it with me.

File: R/otp-route.R
Comment (mem48) [reply R/otp-route.R]: I know that POSIXct is a little awkward, but it does ensure that we get properly formatted dates. Otherwise, there would be a wide range of formats that would need to be supported or checked.


OTP doesn't seem to support routing in the past, so I would not expect a 2017 date to work

File: R/otp-route.R
Comment (mem48) [reply R/otp-route.R:109]: Yes, Google invented it http://www.danmandle.com/blog/what-is-an-encoded-polyline/

File: R/otp-route.R
Comment (marcusyoung) [reply R/otp-route.R]: @mem48 OTP will route based on the schedules provided. If they are in the past then it will route on them.  The schedules in my test instance covered that period and worked in the GUI. But the format must be MM-DD-YYYY in the API request. You can see that in the GUI URL and also see the effect of altering it to YYYY-MM-DD.

File: R/otp-route.R
Comment (marcusyoung) [reply R/otp-route.R:109]: Oh right. Interesting. I didn't realise OTP used them.

File: R/otp-route.R
Comment (marcusyoung) [reply R/otp-route.R]: OTP only supports one format though. So as long as the documentation specified:

time as "13:00:00" and date as "MM-DD-YYYY" we would only have to check for that?

File: vignettes/advanced_features.Rmd
Comment (marcusyoung) [reply vignettes/advanced_features.Rmd:63]: Actually, given the other issue regarding date format mention above, I suspect this is actually the WALK isochrone for Manchester Airport as no Transit travel would have been possible. But WALK won't return a proper valid isochrone, thus all the holes.

File: R/otp-route.R
Comment (marcusyoung) [reply R/otp-route.R]: Obviously the date of travel specified in the request must match the transit service validity period (as provided in calendar.txt in the GTFS feed).

Comment (marcusyoung) [reply R/otp-route.R]: Actually this is one of the great features of OTP (compared to using an online trip planner that will only ever have the current schedules) , you can provide it with old or future data for research. E.g. look at how accessibility has changed due to reduction in bus services over the past 25 years ...

Comment (mem48) [reply R/otp-route.R]: I've adjusted the date and time formatting passed to the API, which I think is part of the problem. I can not get historical routing, as far back as the GTFS file goes. Also, are you using the GTFS in the package or your tutorial one? I made some modifications to the one in the package.

I still would prefer the have POSIXct inputs, but perhaps the vignettes can highlight an easy way to make dates e.g. `date_time = lubridate::ymd_hms("2018-04-07 12:15:00")`

File: R/otp-geocode.R
Comment (mem48) [reply R/otp-geocode.R:28]: fixed, and switched to checkmate checks

File: R/otp-route.R
Comment (marcusyoung) [reply R/otp-route.R]: I was using my tutorial GTFS feed for testing. 
That method of making the date is certainly more streamlined!

**Comment by marcusyoung** [comment]
Replaced by pr17

**Pull request closed without merging**
